### PR TITLE
Model Updates/Graph editor maintenance

### PR DESCRIPTION
**Pull request by Sheyne**
- Restructures the internal model. 
    - Graph editor now keeps its own model of the graph
    - Creates main-graph.ts which syncs `CoreElements` with the graph-editor's model
 - Adds support for the new version of sinap-core.
 - Fixes reusing plugin contexts
 - Reimplements saving/loading in a future proof manner

### CODE REVIEW

**Comment by ilargierdi** [comment]
Related: #103 and #41

File: app/components/graph-editor/defaults.ts [hunk 1]
```diff
@@ -99,20 +97,17 @@ export const NUDGE: number = 3;
...
-export const NODE_DEFAULTS: DrawableNode = {
-    position: { x: 0, y: 0 },
-    label: "",
-    shape: "circle",
-    color: "#fff200",
-    borderColor: "#000",
-    borderStyle: "solid",
-    borderWidth: 1
+ */
+export const NODE_PROPERTIES = {
```
Comment (slaymaker1907) [inline app/components/graph-editor/defaults.ts]: I assume that this are get to avoid mutation, but that seems kind of silly to me. I think it would be more readable as a JSON object constant.

File: app/components/graph-editor/drawable-element.ts [hunk 2]
```diff
@@ -0,0 +1,275 @@
+// File: drawable-element.ts
+// Created by: CJ Dimaano
+// Date created: February 4, 2017
+
+
+import { FONT_SIZE } from "./defaults";
+import { GraphEditorCanvas, point, rect, size } from "./graph-editor-canvas";
+import { Drawable } from "./drawable";
+import { DrawableGraph } from "./drawable-graph";
+import {
```
Comment (slaymaker1907) [inline app/components/graph-editor/drawable-element.ts]: We should decide on multline imports or single line imports to have consistent style.

File: app/components/graph-editor/drawable-element.ts [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    Default,
+    Dragging,
+    Hovered
+};
+
+/**
+ * DrawableElement  
+ *   Common properties shared between drawable edges and nodes.
+ */
+export abstract class DrawableElement extends Drawable {
```
Comment (slaymaker1907) [inline app/components/graph-editor/drawable-element.ts]: This seems pretty complicated. Would it be possible to avoid an abstract class by pulling this out into an interface and a putting the logic into a common class/function?

File: app/components/graph-editor/events.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
+// File: events.ts
+// Created by: CJ Dimaano
+// Date created: February 7, 2017
+
+
+export type Listener<A extends EventArgs> = (evt?: A) => void;
+
+export type PropertyChangedEventListener<T> = Listener<PropertyChangedEventArgs<T>>;
+
+export class EventArgs {
```
Comment (slaymaker1907) [inline app/components/graph-editor/events.ts]: Is there a reason why this is a class instead of an interface?

File: app/components/graph-editor/events.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    constructor(public readonly source: any) { }
+}
+
+export class CancellableEventArgs extends EventArgs {
```
Comment (slaymaker1907) [inline app/components/graph-editor/events.ts]: Same thing about why is this a class instead of an interface?

File: app/components/graph-editor/events.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        public readonly curr: T
+    ) {
+        super(source);
+    }
+}
+
+export class EventEmitter<A extends EventArgs, L extends Listener<A>> {
+    protected listeners: Set<L>
+    = new Set<L>();
+    addListener(l: L) {
```
Comment (slaymaker1907) [inline app/components/graph-editor/events.ts]: This parameter name looks like the number one and should be made more readable.

File: app/components/graph-editor/events.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    emit(args: A) {
+        this.listeners.forEach(v => v(args));
+    }
+}
+
+export class CancellableEventEmitter<A extends CancellableEventArgs, L extends Listener<A>>
+    extends EventEmitter<A, L> {
+    emit(args: A) {
+        for (const l of this.listeners) {
+            l(args);
```
Comment (slaymaker1907) [inline app/components/graph-editor/events.ts]: This parameter should also be renamed for readability.

File: app/models/double-map.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
+
+
+/**
+ * Keep two dictionaries in sync, different efficient maps to the same data
+ */
+export class DoubleMap<A, B, C> {
```
Comment (slaymaker1907) [inline app/models/double-map.ts]: If this is only used one place, then make it non-generic since generic code is inherently more complicated.

File: app/models/main-graph.ts [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+ * 
+ * Calls to get/set will update both. It also conatins a 
+ * `proxy` field. Reads and writes to this will intelligently update 
+ * the core element and the drawable. For any of these uses, if 
+ * the backer object (core/drawable) references another core or drawable
+ * this will return a `BridgingProxy` in its place. Sets will intellegently
+ * sync the backends even if the set `v` is a `BridgingProxy`
+ */
+export class BridgingProxy {
+    proxy: { [a: string]: any };
```
Comment (slaymaker1907) [inline app/models/main-graph.ts]: We should have a mappable interface or something for this for readability.

File: app/models/main-graph.ts [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+
+    set(k: PropertyKey, v: any, updateDrawable = true) {
+        let drawableValue = v;
+        let coreValue = v;
+
+        // If v is a drawable element or a core element,
+        // then when it gets set in the various proxied data
+        // structures, set it differently. 
+
+        if (v instanceof BridgingProxy) {
```
Comment (slaymaker1907) [inline app/models/main-graph.ts]: Could we change BridgingProxy, potentially adding in a second class, so that we don't have to do instanceof checks?

File: app/models/main-graph.ts [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+        const coreValue = this.core.data[k];
+        // if this is something that should be proxied
+        const proxiedValue = this.graph.bridges.getB(coreValue);
+        // return the proxy. That is, any changes made to the returned 
+        // object should propogate to both backers
+        return proxiedValue !== undefined ? proxiedValue : coreValue;
+    }
+}
+
+function drawableFromAny(a: any, bridges: DoubleMap<Drawable, CoreElement, BridgingProxy>) {
```
Comment (slaymaker1907) [inline app/models/main-graph.ts]: This really seems like it should be a method on bridges.

File: app/models/main-graph.ts [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+export class UndoableAdd {
+    constructor(public bridge: BridgingProxy) { }
+}
+
+export class UndoableDelete {
+    constructor(public bridge: BridgingProxy) { }
+}
+
+// TODO: changes can also be adds or deletes
+export class UndoableChange {
```
Comment (slaymaker1907) [inline app/models/main-graph.ts]: What if we just made this an interface with two methods, undo and redo?

File: app/models/main-graph.ts [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+    constructor(public target: BridgingProxy,
+        public key: PropertyKey,
+        public oldValue: any,
+        public newValue: any) { }
+}
+
+export class MainGraph {
```
Comment (slaymaker1907) [inline app/models/main-graph.ts]: This should be renamed to GraphController.

File: app/models/main-graph.ts [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+    drawable: DrawableGraph;
+    activeNodeType: string;
+    activeEdgeType: string;
+    public changed = new EventEmitter<UndoableEvent>();
+
+    private selectedElements: Set<BridgingProxy>;
+
+    public bridges = new DoubleMap<Drawable, CoreElement, BridgingProxy>();
+
+    constructor(public core: CoreModel, public plugin: Plugin) {
```
Comment (slaymaker1907) [inline app/models/main-graph.ts]: It seems like these should be readonly or have an explicit setter.

File: app/services/plugin.service.ts [hunk 15]
```diff
@@ -1,236 +1,61 @@
...
-    }
-    edgePluginData(kind: string) {
-        const type = this.definitions.edges.get(kind);
-        if (!type) {
-            throw "type not found";
-        }
-        return new PluginPropertyData(kind, type);
-    };
+export declare class Program {
+    constructor(any: SerialJSO);
```
Comment (slaymaker1907) [inline app/services/plugin.service.ts]: We should rename this to avoid confusing it with the type any.

File: app/services/plugin.service.ts [hunk 16]
```diff
@@ -1,236 +1,61 @@
...
+    private plugins = new Map<string, Plugin>();
+    private programs = new Map<Plugin, Promise<StubContext>>();
+    private getResults: Script;
+    private addGraph: Script;
     // TODO: load from somewhere
-    private pluginKinds = new Map([[MagicConstants.DFA_PLUGIN_KIND, { definitions: "./dfa-definition.sinapdef", interpreter: "./build/plugins/dfa-interpreter.js" }]])
+    private pluginKinds = new Map([[MagicConstants.DFA_PLUGIN_KIND, "./plugins/dfa-interpreter.ts"]])
 
-    constructor( @Inject(LocalFileService) private fileService: FileService,
+    constructor( @Inject(LocalFileService) private fileService: LocalFileService,
```
Comment (slaymaker1907) [inline app/services/plugin.service.ts]: Why was this changed to use LocalFileService instead of FileService?

File: app/services/plugin.service.ts [hunk 17]
```diff
@@ -1,236 +1,61 @@
...
-        if (loadedPlugin) {
-            return loadedPlugin;
+    public getPlugin(kind: string) {
+        let plugin = this.plugins.get(kind);
+        if (plugin) {
+            return plugin;
+        }
+        const fileName = this.pluginKinds.get(kind);
+        if (!fileName) {
+            throw "Plugin not installed";
```
Comment (slaymaker1907) [inline app/services/plugin.service.ts]: Make sure to specify which plugin kind in the error message.

File: app/services/plugin.service.ts [hunk 18]
```diff
@@ -1,236 +1,61 @@
...
-            }
-            if (value.subtype(Type.Edge)) {
-                edges.push([name, value as ClassMetaType]);
-            }
-            if (value.subtype(Type.Graph)) {
-                graphs.push([name, value as ClassMetaType]);
-            }
+    private getProgramContext(plugin: Plugin) {
+        let contextPromise = this.programs.get(plugin);
+        if (contextPromise === undefined) {
```
Comment (slaymaker1907) [inline app/services/plugin.service.ts]: Change this to just be if(contextPromise) to avoid weird issues with null vs. undefined.

File: plugins/dfa-interpreter.ts [hunk 19]
```diff
@@ -1,207 +1,102 @@
...
-        this.nodes = serialGraph.nodes.map((oldNode: any) => {
-            const result: Node = {
-                label: oldNode.drawableProperties.Label,
-                parents: [],
-                children: []
-            };
-            transferProperties(oldNode, result);
-            return result;
-        });
+export type Nodes = DFANode
```
Comment (slaymaker1907) [inline plugins/dfa-interpreter.ts]: I think these should end with semicolons.

File: plugins/dfa-interpreter.ts [hunk 20]
```diff
@@ -1,207 +1,102 @@
...
-}
-
-interface PropertyObject {
-    [propName: string]: any;
-}
-
-interface Element extends PropertyObject {
+export class DFANode {
+    isStartState: boolean;
+    /** Accept State */
```
Comment (slaymaker1907) [inline plugins/dfa-interpreter.ts]: It doesn't seem like we need this renamed?

File: plugins/dfa-interpreter.ts [hunk 21]
```diff
@@ -1,207 +1,102 @@
...
+
+        if (node.isAcceptState) {
+            accepts.add(node);
         }
-        let symbols = node.children.map((edge: Edge) => edge.label);
-        let uniqueSymbols = new Set(symbols);
-        if (symbols.length !== uniqueSymbols.size) {
-            throw "This interpreter does not handle NFAs. Non-unique edge label detected.";
+
+        if (node.children) {
```
Comment (slaymaker1907) [inline plugins/dfa-interpreter.ts]: children should never be null and should just be an empty list if there are no children.

File: app/components/graph-editor/README.md [hunk 22]
```diff
@@ -1,6 +1,98 @@
-_Last updated: January 31, 2017_
-
-# Notes
+_Last updated: February 11, 2017_
+
+# Remarks
+The graph editor component is a custom Angular2 component that is intended to be
```
Comment (ilargierdi) [inline app/components/graph-editor/README.md]: Maybe in the future, but for now, this is what it is.

File: app/components/graph-editor/drawable-element.ts (hunk 3 above)
Comment (ilargierdi) [inline app/components/graph-editor/drawable-element.ts]: I'm not sure how that would be profoundly different? Other than having an extra type declared somewhere.

File: app/components/graph-editor/defaults.ts (hunk 1 above)
Comment (ilargierdi) [inline app/components/graph-editor/defaults.ts]: I agree. Currently this entire file is a stand-in for what should be a global user settings file, except that it currently only contains stuff related to the graph editor.

File: app/models/main-graph.ts (hunk 9 above)
Comment (Sheyne) [inline app/models/main-graph.ts]: What?

File: app/models/main-graph.ts (hunk 10 above)
Comment (Sheyne) [inline app/models/main-graph.ts]: It would be very complicated.

File: app/components/graph-editor/events.ts (hunk 4 above)
Comment (ilargierdi) [inline app/components/graph-editor/events.ts]: Is there a compelling reason why it should be an interface instead of a class?

File: app/models/main-graph.ts (hunk 12 above)
Comment (Sheyne) [inline app/models/main-graph.ts]: But Daniel needs the details of the change. You could make it also conform to the interface, but this is more data driven rather than encoding the logic of undoing the change into the event.

File: app/models/main-graph.ts (hunk 14 above)
Comment (Sheyne) [inline app/models/main-graph.ts]: Readonly.

File: app/components/graph-editor/README.md (hunk 22 above)
Comment (RighteousRaichu) [inline app/components/graph-editor/README.md]: I think the sentiment is it should be decoupled such that we could move it to its own package (and it is). But that currently moving it wouldn't provide a tangible benefit towards completing Beta.

File: app/components/graph-editor/drawable-element.ts (hunk 2 above)
Comment (RighteousRaichu) [inline app/components/graph-editor/drawable-element.ts]: I think we should focus on readability. Short imports shouldn't need two lines, long imports can be too long if we don't break them up.

File: app/models/main-graph.ts (hunk 12 above)
Comment (RighteousRaichu) [inline app/models/main-graph.ts]: I agree with Sheyne. The changes are useful beyond Undo/Redo. The current implementation is the most workable IMO.

File: app/services/plugin.service.ts (hunk 18 above)
Comment (RighteousRaichu) [inline app/services/plugin.service.ts]: I agree. In general, I prefer `if (value)` checks for checking if a value is defined and explicitly checking for booleans. (ie `if (value == true)`)

File: plugins/dfa-interpreter.ts (hunk 20 above)
Comment (RighteousRaichu) [inline plugins/dfa-interpreter.ts]: @Sheyne why is the naming like this? Did you just wanna try out both ways or having a pretty name?

File: plugins/dfa-interpreter.ts (hunk 21 above)
Comment (RighteousRaichu) [inline plugins/dfa-interpreter.ts]: I made lots of changes to the interpreter just to make it work more as expected. In the general case we have #107 to make sure the values sent to the interpreter match what it's expecting.

File: app/components/graph-editor/defaults.ts (hunk 1 above)
Comment (RighteousRaichu) [inline app/components/graph-editor/defaults.ts]: This probably won't happen until after Beta, but I imagine the settings window could have some behavior where each component that wants settings can have their say. So the settings window would have any number of sections, and the IDE could have what it wants, the graph editor could have what it wants, the plugin service could have what it wants, etc...

File: app/components/properties-panel/properties-panel.component.html [hunk 23]
```diff
@@ -1,77 +1,41 @@
...
-                    <td class="label">{{pair[0]}}</td>
-                    <td [ngSwitch]="pair[1].kind">
-                        <div *ngSwitchCase="'Number'">
-                            <input [ngModel]="selectedEntity[group[1]].get(pair[0])" (ngModelChange)="doChange(selectedEntity, group[1], pair[0], $event)" type="number" 
-                                class="number-input"/>
+    <div *ngIf="!isEmpty">
+        <table *ngFor="let name of fieldNames" class="group-table">
+            <template [ngIf]="fields[name]">
+                <tr><th colspan="2">{{name}}</th></tr>
+                <tr *ngFor="let triple of fields[name]">
```
Comment (RighteousRaichu) [inline app/components/properties-panel/properties-panel.component.html]: Unless you have a better idea, @Sheyne, I'm going to make a function that wraps up triple in a object with named properties (ie `let triple of wrap(fields[name])` -> `triple.name`, `triple.path`, etc...) for readabilities sake.

File: app/components/properties-panel/properties-panel.component.ts [hunk 24]
```diff
@@ -3,90 +3,49 @@
...
-
-    private isEmpty() {
-        if (this.selectedEntity) {
-            for (let group of this.groups) {
-                let g = this.selectedEntity[group[1]];
-                if (g.properties.length > 0) {
-                    return false;
-                }
-            }
+    isEmpty = true;
```
Comment (RighteousRaichu) [inline app/components/properties-panel/properties-panel.component.ts]: We should be more careful in general about marking stuff as private.

File: app/components/graph-editor/README.md (hunk 22 above)
Comment (slaymaker1907) [inline app/components/graph-editor/README.md]: Sounds good. I mostly just wanted to know what the overall plan for that was.

File: app/components/graph-editor/defaults.ts (hunk 1 above)
Comment (slaymaker1907) [inline app/components/graph-editor/defaults.ts]: I'm also planning to support plugin specific settings as well. Not sure exactly how that will work with the overall plugin API, though.

File: app/components/graph-editor/drawable-element.ts (hunk 3 above)
Comment (slaymaker1907) [inline app/components/graph-editor/drawable-element.ts]: I just think that abstract classes usually indicate that there is coupled functionality (since some is provided by the abstract part and some by the subclass) that should be decoupled through composition.

File: app/components/graph-editor/events.ts (hunk 4 above)
Comment (slaymaker1907) [inline app/components/graph-editor/events.ts]: Interfaces usually have less coupling and are easier to extend (i.e. don't have to correspond to inheritance hierarchy).

File: app/models/main-graph.ts (hunk 9 above)
Comment (slaymaker1907) [inline app/models/main-graph.ts]: Something like

```
interface Mappable {
  [a: string]: any;
}
```

File: app/models/main-graph.ts (hunk 10 above)
Comment (slaymaker1907) [inline app/models/main-graph.ts]: Sounds good

File: app/services/plugin.service.ts (hunk 16 above)
Comment (slaymaker1907) [inline app/services/plugin.service.ts]: According to what I read about Angular depedency injection, even though the @Inject is a specific type, we can change it to be something else that fulfills the FileService interface.

File: plugins/dfa-interpreter.ts (hunk 21 above)
Comment (slaymaker1907) [inline plugins/dfa-interpreter.ts]: I don't think that this is related to #107. With the version of the serializer that I wrote, children was explicitly set and was never null. I'm not sure if it was changed so that this could be null or if it is an unnecessary check.

**Comment by RighteousRaichu** [comment]
* DONE Rename to GraphController
* DONE Make sure to specify which plugin kind in the error message.
* DONE Rename any in Program declaration
* DISCUSSED Change this to just be if(contextPromise) to avoid weird issues with null vs. undefined.
* DONE I think these should end with semicolons.
* FUTURE Remove triple[…] in properties-panel.component.html <- Will be done when using custom components to render these.
* DONE Serialize/Deserialize errors through the sandbox, use this to rehookup the status bar at least for compiler errors.
* DONE Mark fields in properties panel component as private
